rust/src/encodings/binary.rs

```diff
@@ -1,3 +1,96 @@
-//! Binary encoding
+//! Var-length Binary Encoding
 //!
+
+use std::io::Result;
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+use arrow_array::types::{ByteArrayType, Int64Type};
+use arrow_array::{Array, ArrayRef, GenericByteArray, Int32Array, Int64Array};
+use arrow_data::ArrayDataBuilder;
+use async_trait::async_trait;
+
+use crate::encodings::Decoder;
+use crate::io::object_reader::ObjectReader;
+
+use super::plain::PlainDecoder;
+
 pub struct BinaryEncoder {}
+
+/// Var-binary encoding decoder.
+pub struct BinaryDecoder<'a, T: ByteArrayType> {
+    reader: &'a ObjectReader<'a>,
+
+    position: usize,
+
+    length: usize,
+
+    phantom: PhantomData<T>,
+}
+
+/// Var-length Binary Decoder
+///
+impl<'a, T: ByteArrayType> BinaryDecoder<'a, T> {
+    /// Create a [BinaryEncoder] to decode one batch.
+    ///
+    ///  - `position`, file position where this batch starts.
+    ///  - `length`, the number of records in this batch.
+    ///
+    /// ## Example
+    ///
+    /// ```rust
+    /// use arrow_array::types::Utf8Type;
+    /// use object_store::path::Path;
+    /// use lance::io::ObjectStore;
+    /// use lance::encodings::binary::BinaryDecoder;
+    ///
+    /// async {
+    ///     let object_store = ObjectStore::new(":memory:").unwrap();
+    ///     let path = Path::from("/data.lance");
+    ///     let reader = object_store.open(&path).await.unwrap();
+    ///     let string_decoder = BinaryDecoder::<Utf8Type>::new(&reader, 100, 1024);
+    /// };
+    /// ```
+    pub fn new(reader: &'a ObjectReader, position: usize, length: usize) -> Self {
+        Self {
+            reader,
+            position,
+            length,
+            phantom: PhantomData,
+        }
+    }
+}
+
+#[async_trait]
+impl<'a, T: ByteArrayType> Decoder for BinaryDecoder<'a, T> {
+    async fn decode(&self) -> Result<ArrayRef> {
+        let position_decoder =
+            PlainDecoder::<Int64Type>::new(self.reader, self.position, self.length + 1)?;
+        let positions = position_decoder.decode().await?;
+        let int64_positions = positions.as_any().downcast_ref::<Int64Array>().unwrap();
+
+        let start_position = int64_positions.value(0);
+        let offset_arr = Int32Array::from(
```

Comment:
how come this down casting doesn't cause overflow issues?

Reply:
So `int64` array is what `lance spec` on disk. But for `utf8/binary` (not large_utf8/large_binary), it's offset array is 32bit.  Overflow should not happen if we write data correctly on the writer side.